### PR TITLE
[deprecated]Move search index to .recovery directory when database is deleted

### PR DESCRIPTION
**Pull request by jiangphcn**
## Overview
Before change, the search index directories and files were directly deleted when database was deleted. This PR provides way of moving search index to directory like `/srv/search_index/.recovery/shard/60000000-7fffffff/<dbname.ts>/e66df316792ab411705e2741bba44371` when the corresponding database was deleted and "enable_database_recovery" configuration item is set to true. This allows search index files to be re-used if database is recovered. 

## Testing recommendations
```
Running com.cloudant.clouseau.IndexCleanupServiceSpec
Running com.cloudant.clouseau.IndexCleanupServiceSpec
2017-07-28 17:12:06 clouseau.cleanup [INFO] Moving 'foo' to '.recovery/foo'
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.02 sec
Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.033 sec

Results :

Tests run: 85, Failures: 0, Errors: 0, Skipped: 0

[INFO]
[INFO] --- maven-jar-plugin:2.2:jar (default-jar) @ clouseau ---
[INFO]
[INFO] --- maven-assembly-plugin:2.3:single (default) @ clouseau ---
[INFO] Reading assembly descriptor: src/main/assembly/distribution.xml
[INFO] Building zip: /Users/jiangph/couchdb/labs-clouseau3/clouseau/target/clouseau-2.10.0-SNAPSHOT.zip
[INFO] Building tar: /Users/jiangph/couchdb/labs-clouseau3/clouseau/target/clouseau-2.10.0-SNAPSHOT.tar.gz
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 24.148 s
[INFO] Finished at: 2017-07-28T17:12:10+08:00
[INFO] Final Memory: 15M/299M
[INFO] -----------------------------------------------------------------------
```

## GitHub issue number
Bugzid: 86318

## Related Pull Requests

https://github.com/cloudant-labs/clouseau/pull/15
https://github.com/cloudant/chef-repo/pull/7456
## Checklist

- [X] Code is written and works correctly;
- [X] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

File: src/dreyfus_index_manager.erl
Comment (rnewson) [inline src/dreyfus_index_manager.erl]: make multiple clauses instead.

File: src/dreyfus_index_manager.erl
Comment (rnewson) [inline src/dreyfus_index_manager.erl]: `RecoveryEnabled` would be a better name, we're not 'doing' recovery here, we're just enabling a future recovery.

**Comment by jiangphcn** [comment]
Thanks so much for your review, Bob. I addressed them and would you please take a look?

**Comment by rnewson** [comment]
These changes look good but don't merge yet as this needs adjusting given the comments on the related clouseau PR.

**Comment by jiangphcn** [comment]
Yes, need to align with the change of clouseau. I modified codes to move calculation of destination directory for recovering index to dreyfus/erlang part. Could you please take a look? Thanks

File: src/dreyfus_index_manager.erl
Comment (rnewson) [inline src/dreyfus_index_manager.erl:91]: This doesn't match the gen_server:cast above and so it doesn't work. When I review code I am assuming that you have run through a successful move, but that doesn't seem possible with this problem.

File: src/dreyfus_index_manager.erl
Comment (rnewson) [inline src/dreyfus_index_manager.erl:130]: same problem here

**Review by rnewson — CHANGES_REQUESTED** [review]

File: src/dreyfus_index_manager.erl
Comment (jiangphcn) [reply src/dreyfus_index_manager.erl:91]: Yes, I run a successful move before having this commit. It looks that `gen_server:cast(?MODULE, {move, DbName})` matches with `handle_cast({move, DbName}, State)`.

File: src/dreyfus_index_manager.erl
Comment (rnewson) [reply src/dreyfus_index_manager.erl:130]: I'm also wrong here again, sorry.

File: src/dreyfus_index_manager.erl
Comment (rnewson) [reply src/dreyfus_index_manager.erl:91]: I see it, I read it backward.

File: src/dreyfus_index_manager.erl
Comment (rnewson) [inline src/dreyfus_index_manager.erl]: the recovery dir should be specified as an absolute path. I think that needs to happen in the dbcore chef cookbook. It should be under the `[:clouseau][:index_dir]` directory.

**Comment by rnewson** [comment]
the move Src Dest code looks fine to me. The next step is to figure out how we'll deploy the change. We need to get an absolute path from the config, so please make a chef-repo PR with that. Once that exists we can ensure all three PR's will work together before merging.

**Comment by jiangphcn** [comment]
Okay, issued the third PR to make directories to be configurable:
https://github.com/cloudant/chef-repo/pull/7456

**Comment by rnewson** [comment]
+1

**Pull request closed without merging**
